src/View/AbstractDirectoryView.vala

```diff
@@ -1834,364 +1837,396 @@ namespace FM {
+                        menu.add (rename_menuitem);
                     }
 
-                    menu.append_section (null, builder.get_object ("rename") as GLib.MenuModel);
-                }
-
-                if (common_actions.get_action_enabled ("bookmark")) {
                     /* Do  not offer to bookmark if location is already bookmarked */
-                    if (window.can_bookmark_uri (selected_files.data.uri)) {
-                        menu.append_section (null, builder.get_object ("bookmark") as GLib.MenuModel);
+                    if (common_actions.get_action_enabled ("bookmark") && window.can_bookmark_uri (selected_files.data.uri)) {
+                        menu.add (bookmark_menuitem);
                     }
-                }
-                menu.append_section (null, builder.get_object ("properties") as GLib.MenuModel);
-            }
 
-            if (menu.get_n_items () > 0) {
-                return menu as MenuModel;
+                    menu.add (new Gtk.SeparatorMenuItem ());
+                    menu.add (properties_menuitem);
+                }
             } else {
-                return null;
-            }
-        }
+                var show_hidden_menuitem = new Gtk.CheckMenuItem.with_label (_("Show Hidden Files"));
+                show_hidden_menuitem.action_name = "background.show-hidden";
+
+                var show_remote_thumbnails_menuitem = new Gtk.CheckMenuItem.with_label (_("Show Remote Thumbnails"));
+                show_remote_thumbnails_menuitem.action_name = "background.show-remote-thumbnails";
 
-        private GLib.MenuModel? build_menu_background (ref Gtk.Builder builder, bool in_trash, bool in_recent) {
-            var menu = new GLib.Menu ();
+                var hide_local_thumbnails_menuitem = new Gtk.CheckMenuItem.with_label (_("Hide Thumbnails"));
+                hide_local_thumbnails_menuitem.action_name = "background.hide-local-thumbnails";
 
-            if (in_trash) {
-                if (clipboard != null && clipboard.has_cutted_file (null) ) {
-                    menu.append_section (null, builder.get_object ("paste") as GLib.MenuModel);
-                    return menu as MenuModel;
+                if (in_trash) {
+                    if (clipboard != null && clipboard.has_cutted_file (null)) {
+                        menu.add (paste_menuitem);
+                    }
+                } else if (in_recent) {
+                    menu.add (new Gtk.SeparatorMenuItem ());
+                    menu.add (new SortSubMenuItem ());
+                    menu.add (new Gtk.SeparatorMenuItem ());
+                    menu.add (show_hidden_menuitem);
+
+                    if (slot.directory.is_local) {
+                        menu.add (hide_local_thumbnails_menuitem);
+                    } else if (slot.directory.can_open_files) {
+                        menu.add (show_remote_thumbnails_menuitem);
```

Comment:
Is it reasonable to think the recent folder will ever be remote?

Reply:
I'm not familiar enough with how that's implemented, but I don't think the entire directory could be remote… it's a virtual directory. I guess maybe theoretically items inside of it could be remote?

Reply:
Yeah this is more a question for @jeremypw. My guess is that this is unreachable code but I'd like to ask before removing it